rules/match.go

```diff
@@ -21,46 +21,52 @@
 package rules
 
 import (
+	"github.com/m3db/m3collector/metric"
 	"github.com/m3db/m3metrics/rules"
 	"github.com/m3db/m3x/errors"
 )
 
 // Matcher matches rules against metric IDs.
 type Matcher interface {
 	// Match matches rules against metric ID and returns the match result.
-	Match(id []byte) rules.MatchResult
+	Match(id metric.ID) rules.MatchResult
 
 	// Close closes the matcher.
 	Close() error
 }
 
 type matcher struct {
-	opts        Options
-	namespaceFn NamespaceFn
+	opts             Options
+	namespaceTag     string
+	defaultNamespace string
 
-	namespaces  *namespaces
-	cache       Cache
+	namespaces *namespaces
+	cache      Cache
 }
 
 // NewMatcher creates a new rule matcher.
 func NewMatcher(cache Cache, opts Options) (Matcher, error) {
 	key := opts.NamespacesKey()
 	namespaces := newNamespaces(key, cache, opts)
-	if err := namespaces.Start(); err != nil {
+	if err := namespaces.Watch(); err != nil {
 		return nil, err
 	}
 
 	return &matcher{
-		opts:        opts,
-		namespaceFn: opts.NamespaceFn(),
-		namespaces:  namespaces,
-		cache:       cache,
+		opts:             opts,
+		namespaceTag:     opts.NamespaceTag(),
+		defaultNamespace: opts.DefaultNamespace(),
+		namespaces:       namespaces,
+		cache:            cache,
 	}, nil
 }
 
-func (m *matcher) Match(id []byte) rules.MatchResult {
-	namespace := m.namespaceFn(id)
-	return m.cache.Get(namespace, id)
+func (m *matcher) Match(id metric.ID) rules.MatchResult {
+	ns, found := id.TagValue(m.namespaceTag)
+	if !found {
+		return m.cache.Match(m.defaultNamespace, id.Bytes())
+	}
+	return m.cache.Match(string(ns), id.Bytes())
```

Comment:
Haha, ah yeah. Hm, is it worth making namespace all `[]byte` everywhere too?  Or at least in a follow up PR with a comment above this?